feat: Indexing BOPIS Data (storeAvailability) for Variant Record Model

htuzel opened the pull request:
## What is Changed

- Added 'storeAvailability' to default variant attributes in algoliaProductConfig
- Updated algoliaLocalizedProduct to retrieve and index store inventory information
- Implemented a new attribute handler for storeAvailability that checks product inventory across stores
- Supports optional indexing of store-level product availability based on site preference

## How to Test

- Install BOPIS cartridge.
- Add `storeAvailability` in Additional Product Attributes and run the `AlgoliaProductIndex_v2` job.
- Need to see storeAvailability array in index

sbellone commented on cartridges/int_algolia/cartridge/scripts/algolia/model/algoliaLocalizedProduct.js:
```diff
@@ -30,6 +30,28 @@ try {
+    var storeIt = SystemObjectMgr.getAllSystemObjects('Store');
+    while (storeIt.hasNext()) {
+        var store = storeIt.next();
+        var storeInventoryListId = store.inventoryListID;
+        if (storeInventoryListId) {
+            var storeInventory = ProductInventoryMgr.getInventoryList(storeInventoryListId);
```
You don't need that, you can access the inventoryList directly from the `Store` object: `store.inventoryList`

htuzel replied:
good point! Refactored!

sbellone commented on cartridges/int_algolia/cartridge/scripts/algolia/model/algoliaLocalizedProduct.js:
```diff
@@ -30,6 +30,28 @@ try {
 
 const ALGOLIA_IN_STOCK_THRESHOLD = algoliaData.getPreference('InStockThreshold');
 const INDEX_OUT_OF_STOCK = algoliaData.getPreference('IndexOutOfStock');
+const ATTRIBUTE_LIST = algoliaData.getSetOfArray('AdditionalAttributes');
+const stores = [];
+var SystemObjectMgr;
+var ProductInventoryMgr;
+
+if (ATTRIBUTE_LIST.indexOf('storeAvailability') !== -1) {
```
Can this be moved outside this model? As you saw previously, accessing preferences and doing global things makes it harder to test.
We can imagine the jobs building the list of `Store` with an inventory list and passing it to the model:

```js
var storesList = [];
if (attributesToSend.indexOf('storeAvailability') !== -1) {
  var storeIt = SystemObjectMgr.getAllSystemObjects('Store');
  while (storeIt.hasNext()) {
    var store = storeIt.next();
    if (store.inventoryList) {
      storesList.push(store);
    }
  }
}
// ...
new AlgoliaLocalizedProduct({ storesList: storeList, ...
```

htuzel replied:
Sure, I agree with you regarding it causing tests to be difficult, but we are calling AlgoliaLocalizedProduct on different files and places. In this case, we will write duplicated code, and I think this approach is cleaner.

sbellone replied:
Fair enough

sbellone commented on cartridges/int_algolia/cartridge/scripts/algolia/model/algoliaLocalizedProduct.js:
```diff
@@ -466,6 +488,25 @@ var aggregatedValueHandlers = {
+                return storeArray;
+            }
+        }
+        return undefined;
```
Let's index an empty array by default so it's clearer that the field was indexed but no store had availability.
```suggestion
        return storeArray;
```

sbellone reviewed (commented):
It works well, I'd just prefer to keep the `algoliaProductRecord` model simple.

htuzel commented:
I refactored some parts according to your comments. Also, added unit tests. Lastly, please let me know what is the best solution for the structure here for `AlgoliaLocalizedProduct` (I mean passing stores as parameter or calculating it on model)

htuzel commented on test/unit/int_algolia/scripts/algolia/model/storeAvailability.test.js:
Moved here

sbellone commented on test/unit/int_algolia/scripts/algolia/model/algoliaLocalizedProduct.test.js:
```diff
@@ -64,6 +64,76 @@ jest.mock('*/cartridge/scripts/algolia/customization/productModelCustomizer', ()
+    return {
+        getAllSystemObjects: jest.fn().mockImplementation((objType) => {
+            if (objType === 'Store') {
+                const storeIterator = {
+                    hasNext: function() {
```
I just noticed that: we have a mock for `Collection` which already implements an iterator, let's use it: https://github.com/algolia/algoliasearch-sfcc-b2c/blob/15481bb26375883fbdf5d335a9136d546cfb4c29/test/mocks/dw/util/Collection.js
